[BEAM-851] Identify if pipeline is in streaming mode

iemejia opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

iemejia commented:
R: @amitsela @jbonofre

jbonofre commented on runners/spark/pom.xml:
```diff
@@ -34,7 +34,7 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-    <spark.version>1.6.2</spark.version>
+    <spark.version>1.6.3</spark.version>
```
Does the PR really require Spark 1.6.3 update (I don't think so). I would not mix different purpose in the same PR: Spark 1.6.3 upgrade should be in a dedicated PR IMHO.

amitsela replied:
+1 for JB's comment, we should have a separate PR for that.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineOptions.java:
```diff
@@ -23,16 +23,13 @@
  * such as the master address, batch-interval, and other user-related knobs.
  */
 public interface SparkPipelineOptions
-    extends PipelineOptions, StreamingOptions, ApplicationNameOptions {
```
I wouldn't remove the `isStreaming` option, but set it instead.
For one, SDK tests for streaming would require "forcing" streaming on bounded sources, and that it communicated via `PipelineOptions.isStreaming()`

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/utils/SparkTestPipelineOptionsForStreaming.java:
```diff
@@ -28,12 +28,6 @@
  */
 public class SparkTestPipelineOptionsForStreaming extends SparkTestPipelineOptions {
 
-  @Override
```
👍 this can actually go is the `isStreaming()` option will be automagically set 😄 
Please check if other unit tests are explicitly setting streaming, thanks!

iemejia replied:
all of they used the SparkTestPipelineOptionsForStreaming so it can be removed, I hesitated about removing the whole class, because it only calls the method of the father, what do you think ?

amitsela commented:
@iemejia thanks!
See comments and ping me once updated.

iemejia commented:
Ok, rebased with the changes, please check @amitsela

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java:
```diff
@@ -120,13 +121,14 @@ private SparkRunner(SparkPipelineOptions options) {
     try {
       LOG.info("Executing pipeline using the SparkRunner.");
-
-      if (mOptions.isStreaming()) {
```
Why not keep relying on options, and run the detecting visitor before this line ?
BTW, not sure why Flink named this optimizer... I think something like `StreamingPipelineDetector/StreamingModeDetector` is more fitting.

amitsela commented:
Jenkins is failing on checkstyle. Please run `mvn clean install -Prelease` locally to see the issues.

iemejia commented:
Ok sorry for the checkstyle issue, I just fixed it, I also changed the options to be aligned with what you said, the TranslationModeDetector only detects the mode and does not do any extra logic (I suppose this is better if someday we eventually reuse this function between runners).

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java:
```diff
@@ -168,6 +169,62 @@ public EvaluationResult run(Pipeline pipeline) {
+    public void visitPrimitiveTransform(TransformTreeNode node) {
+      if (translationMode.equals(TranslationMode.BATCH)) {
+        Class<? extends PTransform> transformClass = node.getTransform().getClass();
+        if (transformClass == Read.Unbounded.class) {
```
For now, this is indeed the only way to create a streaming pipeline. This might change in the future with `SplittableDoFn`, but that's fine.

amitsela commented:
@iemejia looking good!
My comments are mostly thinking out loud 😉 .

I'm wondering though if you could remove all specific assignments of `options.setStreaming(true)` in the runner ? this would basically test your changes.

jbonofre commented:
I have a test failure on the PR. Let me try a new full build.

iemejia commented:
Actually If I remove this, some of the tests will fail because the way we detect using Read.Unbounded does not match unbounded PCollections created by transforms e.g. CreateStream.fromQueue.
So there are two options:
1. I change back the visitor code to check every input for unboundness (my previous version, notice that the code is not as nice but it will cover event the SDF case).
2. I switch the respective tests to force streaming when needed and let the visitor detec only Unbounded.Read.

amitsela commented:
@iemejia you're right, I forgot about it.
I suggest we `setStreaming(true)` manually where `CreateStream` is used as it is not a proper IO (possibly fix that in another PR).
WDYT ?

iemejia commented:
Ok, added last remarks, I hope it is ok now.

amitsela commented:
LGTM pending tests.
